pkg/kubelet/cm/dra/manager.go

```diff
@@ -175,27 +171,44 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 				return fmt.Errorf("NodePrepareResources failed for claim %s/%s: %s", reqClaim.Namespace, reqClaim.Name, result.Error)
 			}
 
-			claimInfo := claimInfos[types.UID(claimUID)]
-
-			// Add the CDI Devices returned by NodePrepareResources to
-			// the claimInfo object.
-			err = claimInfo.addCDIDevices(pluginName, result.GetCDIDevices())
-			if err != nil {
-				return fmt.Errorf("failed to add CDIDevices to claimInfo %+v: %+v", claimInfo, err)
+			resourceClaim := resourceClaims[types.UID(claimUID)]
+			if !m.cache.contains(resourceClaim.Name, pod.Namespace) {
+				// Grab the allocation.resourceHandles. If there are no
+				// allocation.resourceHandles, create a single resourceHandle with no
+				// content. This will trigger processing of this claim by a single
+				// kubelet plugin whose name matches resourceClaim.Status.DriverName.
+				resourceHandles := resourceClaim.Status.Allocation.ResourceHandles
+				if len(resourceHandles) == 0 {
+					resourceHandles = make([]resourceapi.ResourceHandle, 1)
+				}
+				claimInfo := newClaimInfo(
+					resourceClaim.Status.DriverName,
+					resourceClaim.Spec.ResourceClassName,
+					resourceClaim.UID,
+					resourceClaim.Name,
+					resourceClaim.Namespace,
+					sets.New(string(pod.UID)),
+					resourceHandles,
+					true,
+				)
+				if err = m.cache.add(&claimInfo); err != nil {
+					return fmt.Errorf("failed to cache claimInfo %+v: %w", claimInfo, err)
+				}
```

Comment:
Looking at this again, it feels weird to me to have the creation of this entry happen all the way at the end like this (not to mention that it is done inside a loop, where it is only created the first time around the loop).

I prefer the flow we had originally which (with the new abstractions) would be something like the following at the top of this function:
```
if !m.cache.contains() {
    m.cache.newEntry(claim)
}

if err := m.cache.addPodReference(); err != nil {
    return ...
}

if m.cache.isPrepared() {
    continue
)
```
Where `newEntry()` creates the cache entry up front using the logic shown here, and then there is nothing to do at this point. What purpose did moving the logic here serve in your original refactoring?

Reply:
The reason was to postpone caching new claims until `PrepareResources` GRPC succeeds. Unprepared claimInfo doesn't influence any logic, but creates possibility for a data race.

Reply:
I'd actually do it even later, outside of the loop, just before checkpointing to decrease data race possibility even more. WDYT?

Reply:
I'm not following the discussion in detail but the wording "decrease data race possibility even more" worries me. If we are not sure that the code is free of race conditions, then it should be simplified further until it is simple enough to understand and convince a reader that there are no race conditions.

Reply:
Agreed. Ed and I discussed offline and I have put together a new set of changes that ed will be pulling into his PR today.